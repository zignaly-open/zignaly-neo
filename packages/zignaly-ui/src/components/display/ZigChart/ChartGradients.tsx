import React from "react";
import { ChartGradientColor } from "./types";
import { dark } from "../../../theme";

const ChartGradients = () => (
  <svg style={{ width: 0, height: 0, display: "block" }}>
    <defs>
      <linearGradient id={ChartGradientColor.RedMini} x1="1%" y1="100%" x2="0%" y2="0%">
        <stop offset="0%" stopColor={dark.chart.redMiniGradient[0]} />
        <stop offset="50%" stopColor={dark.chart.redMiniGradient[1]} />
        <stop offset="100%" stopColor={dark.chart.redMiniGradient[2]} />
      </linearGradient>

      <linearGradient id={ChartGradientColor.GreenMini} x1="1%" y1="100%" x2="0%" y2="0%">
        <stop offset="0%" stopColor={dark.chart.greenMiniGradient[0]} />
        <stop offset="50%" stopColor={dark.chart.greenMiniGradient[1]} />
        <stop offset="100%" stopColor={dark.chart.greenMiniGradient[2]} />
      </linearGradient>
      <linearGradient id={ChartGradientColor.RedCard} x1="0" y1="1" x2="0" y2="0">
        <stop offset="0%" stopColor={dark.chart.redCard[0]} stopOpacity="0.76" />
        <stop offset="32%" stopColor={dark.chart.redCard[1]} stopOpacity="0.76" />
        <stop offset="100%" stopColor={dark.chart.redCard[2]} stopOpacity="0.76" />
      </linearGradient>
      <linearGradient id={ChartGradientColor.GreenCard} x1="0" y1="1" x2="0" y2="0">
        <stop offset="0%" stopColor={dark.chart.greenCard[0]} stopOpacity="0.76" />
        <stop offset="32%" stopColor={dark.chart.greenCard[1]} stopOpacity="0.76" />
        <stop offset="53%" stopColor={dark.chart.greenCard[2]} stopOpacity="0.76" />
        <stop offset="100%" stopColor={dark.chart.greenCard[3]} stopOpacity="0.76" />
      </linearGradient>
      <linearGradient id={ChartGradientColor.RedFull} x1="0" y1="1" x2="0" y2="0">
        <stop offset="0%" stopColor={dark.chart.redGradient[0]} />
        <stop offset="100%" stopColor={dark.chart.redGradient[1]} />
      </linearGradient>

      <linearGradient id={ChartGradientColor.GreenFull} x1="0" y1="1" x2="0" y2="0">
        <stop offset="0%" stopColor={dark.chart.greenGradient[0]} />
        <stop offset="100%" stopColor={dark.chart.greenGradient[0]} />
      </linearGradient>
    </defs>
  </svg>
);

export default ChartGradients;
