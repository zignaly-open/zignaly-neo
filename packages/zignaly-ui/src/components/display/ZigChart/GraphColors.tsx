import React from "react";
import { ChartGradientColor, GradientVariant } from "./types";

const GraphColors = ({ variant = "full" }: { variant?: GradientVariant }) => (
  <svg style={{ width: 0, height: 0, display: "block" }}>
    <defs>
      {variant === "short" ? (
        <>
          <linearGradient id={ChartGradientColor.Red} x1="1%" y1="100%" x2="0%" y2="0%">
            <stop offset="0%" stopColor={"rgba(18, 20, 39, 0)"} />
            <stop offset="50%" stopColor={"rgba(21, 21, 57, 0.5)"} />
            <stop offset="100%" stopColor={"rgba(86, 36, 108, 1)"} />
          </linearGradient>

          <linearGradient id={ChartGradientColor.Green} x1="1%" y1="100%" x2="0%" y2="0%">
            <stop offset="0%" stopColor={"rgba(17, 27, 47, 0)"} />
            <stop offset="50%" stopColor={"rgba(22, 41, 67, 0.5)"} />
            <stop offset="100%" stopColor={"rgba(39, 110, 107, 1)"} />
          </linearGradient>
        </>
      ) : (
        <>
          <linearGradient id={ChartGradientColor.RedCard} x1="0" y1="1" x2="0" y2="0">
            {/* <stop offset="0%" stop-color="rgba(46, 0, 28, 0.32)" stop-opacity="0.76" /> */}
            <stop offset="0%" stop-color="rgba(37, 15, 38, 0.38)" stop-opacity="0.76" />
            <stop offset="32%" stop-color="rgba(46, 0, 28, 0.44)" stop-opacity="0.76" />
            <stop offset="100%" stop-color="rgba(204, 57, 147, 0.69)" stop-opacity="0.76" />
          </linearGradient>
          <linearGradient id={ChartGradientColor.GreenCard} x1="0" y1="1" x2="0" y2="0">
            {/* <stop offset="0%" stop-color="rgba(0, 45, 20, 0.74)" stop-opacity="0.76" /> */}
            <stop offset="0%" stop-color="rgba(16, 33, 33, 0.68)" stop-opacity="0.76" />
            <stop offset="32%" stop-color="rgba(0, 45, 20, 0.44)" stop-opacity="0.76" />
            <stop offset="53%" stop-color="rgba(7, 47, 30, 0.44)" stop-opacity="0.76" />
            <stop offset="100%" stop-color="rgba(56, 235, 170, 0.69)" stop-opacity="0.76" />
          </linearGradient>
          <linearGradient id={ChartGradientColor.RedLong} x1="0" y1="1" x2="0" y2="0">
            <stop offset="0%" stop-color="rgba(31, 18, 59, 0.52)" />
            <stop offset="100%" stop-color="rgba(74, 33, 81, 0.69)" />
          </linearGradient>

          <linearGradient id={ChartGradientColor.GreenLong} x1="0" y1="1" x2="0" y2="0">
            <stop offset="0%" stop-color="rgba(18, 33, 59, 0.52)" />
            <stop offset="100%" stop-color="rgba(33, 81, 78, 0.69)" />
          </linearGradient>
        </>
      )}
    </defs>
  </svg>
);

export default GraphColors;
