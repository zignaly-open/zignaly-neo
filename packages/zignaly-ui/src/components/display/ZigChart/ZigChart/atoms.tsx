import React from "react";
import { useTheme } from "@mui/material";
import { VictoryTooltip, VictoryLabel, LineSegment, Point, VictoryTooltipProps } from "victory";

export const ChartTooltip = React.memo(
  ({ color, ...props }: VictoryTooltipProps & { color?: string }) => {
    const theme = useTheme();

    if (!props.text) return null;

    return (
      <g>
        <VictoryTooltip
          {...props}
          orientation="top"
          cornerRadius={8}
          flyoutStyle={{ fill: theme.palette.neutral900, stroke: "none" }}
          pointerOrientation="bottom"
          flyoutPadding={8}
          labelComponent={<VictoryLabel lineHeight={1.5} />}
          style={[
            {
              fill: theme.palette.neutral300,
              fontSize: "14px",
              fontFamily: "Avenir Next",
            },
            {
              fill: theme.palette.neutral200,
              fontSize: "15px",
              fontWeight: 500,
              fontFamily: "Avenir Next",
            },
          ]}
        />
        <LineSegment
          x1={props.x}
          y1={props.y}
          x2={props.x}
          y2={props.height ? props.height - 20 : 0}
          style={{
            stroke: "#3e495b",
          }}
        />
        <Point
          x={props.x}
          y={props.y}
          size={5}
          style={{
            stroke: color ?? "#4c535e",
            fill: "#353234",
          }}
        />
      </g>
    );
  },
  (prevProps, nextProps) => prevProps.x === nextProps.x && prevProps.y === nextProps.y,
);
