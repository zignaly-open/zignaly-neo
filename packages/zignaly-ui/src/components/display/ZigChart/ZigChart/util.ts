import { ChartColors } from "../types";
import { useTheme } from "@mui/material";

export function useChartColor(): ChartColors {
  const theme = useTheme();
  return {
    green: theme.palette.chart.green,
    red: theme.palette.chart.red,
  };
}
