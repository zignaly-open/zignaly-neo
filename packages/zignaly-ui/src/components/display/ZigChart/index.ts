export { default as ZigChartMini } from "./ZigChartMini";
