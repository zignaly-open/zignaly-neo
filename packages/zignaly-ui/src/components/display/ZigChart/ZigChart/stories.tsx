import ZigChart from ".";
import { ComponentMeta, ComponentStory } from "@storybook/react";
import * as React from "react";

export default {
  title: "Display/Chart/ZigChart",
  component: ZigChart,
  argTypes: {
    midLine: {
      control: "radio",
      options: { on: true, off: false },
    },
    height: {
      control: "number",
    },
  },
} as ComponentMeta<typeof ZigChart>;

const Template: ComponentStory<typeof ZigChart> = (args) => <ZigChart {...args} />;

export const ExampleChart = Template.bind({});
ExampleChart.args = {
  data: (() => {
    let value = 100;
    // say "trista"
    return new Array(300).fill(0).map((_, i) => {
      value += Math.random() * 200 - 1;
      return {
        x: "Jun " + i,
        y: value,
      };
    });
  })(),
};

export const ExampleChartWithEvents = Template.bind({});
ExampleChartWithEvents.args = {
  events: [
    { x: 25, label: "Raz raz raz" },
    { x: 145, label: "Eto hardbass" },
  ],
  data: (() => {
    let value = 10;
    // say "trista"
    return new Array(300).fill(0).map((_, i) => {
      value += Math.random() * 20 - 1;
      return {
        x: "Jun " + i,
        y: value,
      };
    });
  })(),
};

export const ExampleChart2 = Template.bind({});
ExampleChart2.args = {
  data: [
    { x: "Jul 1", y: 10 },
    { x: "Jul 2", y: 15 },
    { x: "Jul 3", y: 23 },
    { x: "Jul 4", y: 15 },
    { x: "Jul 5", y: -17 },
    { x: "Jul 6", y: 20 },
    { x: "Jul 7", y: -25 },
  ],
};

export const ExampleChartSlightlyNegative = Template.bind({});
ExampleChartSlightlyNegative.args = {
  data: [
    { x: "Jul 1", y: 10 },
    { x: "Jul 2", y: 15 },
    { x: "Jul 3", y: 23 },
    { x: "Jul 4", y: 15 },
    { x: "Jul 5", y: -1 },
    { x: "Jul 6", y: 20 },
    { x: "Jul 7", y: -2 },
  ],
};

export const ExampleBarChart = Template.bind({});
ExampleBarChart.args = {
  bars: true,
  data: (() => {
    let value = 100;
    // say "trista"
    return new Array(90).fill(0).map((_, i) => {
      value += Math.random() * 200 - 1;
      return {
        x: "Jun " + i,
        y: value,
      };
    });
  })(),
};

export const ExampleBarChartWithEvents = Template.bind({});
ExampleBarChartWithEvents.args = {
  events: [
    { x: 25, label: "Raz raz raz" },
    { x: 145, label: "Eto hardbass" },
  ],
  bars: true,
  data: (() => {
    let value = 10;
    // say "trista"
    return new Array(300).fill(0).map((_, i) => {
      value += (Math.random() - 0.5) * 20;
      return {
        x: "Jun " + i,
        y: value,
      };
    });
  })(),
};

export const ExampleBarChart2 = Template.bind({});
ExampleBarChart2.args = {
  bars: true,
  data: [
    { x: "Jul 1", y: 10 },
    { x: "Jul 2", y: 15 },
    { x: "Jul 3", y: 23 },
    { x: "Jul 4", y: 15 },
    { x: "Jul 5", y: -17 },
    { x: "Jul 6", y: 20 },
    { x: "Jul 7", y: -25 },
  ],
};
