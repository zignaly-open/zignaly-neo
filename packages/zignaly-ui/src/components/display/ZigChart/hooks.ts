import { useMemo } from "react";
import { AxisFormat, ChartGradientColor, GradientVariant } from "./types";
import { useChartColor } from "./ZigChart/util";

const deltaToShowSecondChart = 0.2;

const getGradient = (gradientVariant: GradientVariant, isGreen: boolean) => {
  switch (gradientVariant) {
    case "mini":
      return isGreen ? ChartGradientColor.GreenMini : ChartGradientColor.RedMini;
    case "card":
      return isGreen ? ChartGradientColor.GreenCard : ChartGradientColor.RedCard;
    default:
      return isGreen ? ChartGradientColor.GreenFull : ChartGradientColor.RedFull;
  }
};

/**
 * Get y domain for chart, adjusted to leave space under 0 axis if needed
 */
const getYDomain = (data: AxisFormat[]) => {
  const values = data.map((s) => s.y);
  // Add 0 to min values to show chart under 0 axis
  const ranges = [Math.min(0, ...values), Math.max(...values)];
  if (ranges[0] < 0 && ranges[1] > 0) {
    ranges[0] = Math.min(
      ranges[0],
      (ranges[1] * -1 * deltaToShowSecondChart) / (1 - deltaToShowSecondChart),
    );
  } else if (ranges[1] <= 0) {
    // If all values are negative, we need to force the upper value to be slightly above 0 to
    // avoid the x axis being on top.
    // This doesn't work: https://stackoverflow.com/a/53396988/1494428

    // Calculate the data range, then set the upper limit to a small fraction of that range
    ranges[1] = Math.abs(ranges[1] - ranges[0]) * 0.01;
  }

  return ranges as [number, number];
};

export function useChartData(
  data: AxisFormat[] | number[],
  gradientVariant = "full" as GradientVariant,
  precision?: number,
): {
  data: AxisFormat[];
  color: string;
  gradient: ChartGradientColor;
  yDomain: [number, number];
} {
  const chartColors = useChartColor();
  const [processedData, yDomain] = useMemo(() => {
    const chart = data.map((value, index) => {
      const { x, y, ...rest } = typeof value === "object" ? value : { x: index, y: value };
      return {
        x,
        // Remove extra decimals to normalize range (e.g. avoiding 0 to 0.0001%)
        y: precision ? +y.toFixed(precision) : y,
        // Keep rest of properties that may be useful for tooltip
        ...rest,
      };
    });
    const yDomain = getYDomain(chart);
    return [chart.map((c) => ({ ...c, y0: yDomain[0] })), yDomain];
  }, [data]);

  const firstTimestamp = processedData[0].y;
  const lastTimeStamp = processedData[data.length - 1].y;
  const isGreen = firstTimestamp <= lastTimeStamp;

  return {
    data: processedData,
    color: isGreen ? chartColors.green : chartColors.red,
    gradient: getGradient(gradientVariant, isGreen),
    yDomain,
  };
}
