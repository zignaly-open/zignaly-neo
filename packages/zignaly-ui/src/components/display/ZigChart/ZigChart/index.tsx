import React, { useCallback, useLayoutEffect, useMemo, useReducer, useRef } from "react";
import {
  VictoryArea,
  VictoryAxis,
  VictoryLine,
  VictoryChart,
  VictoryScatter,
  VictoryLabel,
  VictoryBar,
  VictoryVoronoiContainer,
} from "victory";
import { axisStyle, ChartLayoutLarge } from "../styles";
import { ChartColor, ChartLargeProps } from "../types";
import { useChartData } from "../hooks";
// eslint-disable-next-line import/no-extraneous-dependencies
import * as d3Scale from "victory-vendor/d3-scale";
// eslint-disable-next-line import/no-extraneous-dependencies
import { useTheme } from "@mui/material";
import { ChartTooltip } from "./atoms";

const ZigChart = ({
  data,
  yAxisFormatter,
  tooltipFormatter,
  events,
  bars,
  tickCount = 7,
  onlyIntegerTicks,
  chartProps = {},
}: ChartLargeProps) => {
  const theme = useTheme();
  const { data: processedData, color, gradient } = useChartData(data, "full");
  const wrapperRef = useRef<HTMLDivElement>(null);
  const width = wrapperRef?.current?.getBoundingClientRect().width;
  const pureChartWidth = width ? width - 70 - 2 : 0;
  const barChartWidthAdjustedForPadding = Math.min(25, pureChartWidth / (processedData.length + 2));

  // dirty fix for rerender
  const [, forceUpdate] = useReducer((x) => x + 1, 0);
  useLayoutEffect(() => {
    forceUpdate();
  }, [width]);

  const yDomain = useMemo(() => {
    const values = processedData.map((s) => s.y);
    return [Math.min(0, ...values), Math.max(1, ...values)];
  }, [processedData]);

  const getChartLabel = useCallback(
    ({ datum = 0 }: { datum?: number }): string =>
      yAxisFormatter ? yAxisFormatter(datum) : datum.toString(),
    [yAxisFormatter],
  );

  const ticks = d3Scale
    .scaleLinear()
    .domain(yDomain)
    .ticks(tickCount)
    .filter((v) => !onlyIntegerTicks || Number.isInteger(v));

  return (
    <ChartLayoutLarge ref={wrapperRef}>
      {width && (
        <VictoryChart
          containerComponent={
            <VictoryVoronoiContainer
              voronoiDimension="x"
              labels={(point) => tooltipFormatter?.(point.datum) ?? " "}
              labelComponent={<ChartTooltip color={!bars ? color : undefined} />}
              voronoiBlacklist={["eventLine", "scatterText"]}
            />
          }
          {...{
            domain: { y: yDomain as unknown as undefined },
            width: width || 600,
            height: 300,
            domainPadding: { x: 0, y: 1 },
            padding: { left: 35, top: 20, right: 35, bottom: 20 },
          }}
          {...chartProps}
        >
          <VictoryAxis
            tickValues={ticks}
            tickLabelComponent={<VictoryLabel textAnchor="end" text={getChartLabel} />}
            dependentAxis
            style={axisStyle}
            fixLabelOverlap
          />

          {(events || []).map(({ x, label }) => (
            <VictoryScatter
              key={"event-text-" + x}
              width={0}
              data={[{ x, y: yDomain[1] }]}
              labels={[label]}
              size={0}
              name="scatterText"
              labelComponent={
                <VictoryLabel
                  dy={17}
                  labelPlacement="vertical"
                  style={[{ fontSize: 14, fill: theme.palette.neutral500 }]}
                  angle={-90}
                  textAnchor="end"
                />
              }
            />
          ))}

          {(events || []).map(({ x }) => (
            <VictoryLine
              name="eventLine"
              key={"event-line-" + x}
              style={{
                data: { stroke: theme.palette.neutral500, strokeWidth: 0.5 },
              }}
              data={[
                { x, y: yDomain[0] },
                { x, y: yDomain[1] },
              ]}
            />
          ))}

          <VictoryAxis
            tickFormat={() => ""}
            tickLabelComponent={<VictoryLabel />}
            style={axisStyle}
          />

          <VictoryAxis
            offsetY={20}
            tickLabelComponent={<VictoryLabel backgroundPadding={{ top: 5 }} />}
            fixLabelOverlap
            style={
              bars
                ? {
                    ...axisStyle,
                    grid: { stroke: theme.palette.neutral700, strokeDasharray: "3 3" },
                    ticks: { stroke: theme.palette.neutral700, size: 5 },
                  }
                : axisStyle
            }
          />

          {bars ? (
            <VictoryBar
              barRatio={0.9}
              animate={{
                duration: 2000,
                onLoad: { duration: 1000 },
              }}
              barWidth={barChartWidthAdjustedForPadding}
              style={{
                data: {
                  fill: ({ datum }) => {
                    return datum.y > 0 ? ChartColor.Green : ChartColor.Red;
                  },
                  maxWidth: 20,
                  strokeWidth: 0,
                },
              }}
              data={processedData.map((v) => ({ ...v, y0: 0 }))}
            />
          ) : (
            <VictoryArea
              style={{
                data: {
                  fill: `url(#${gradient})`,
                  strokeWidth: 2,
                  stroke: color,
                },
              }}
              data={processedData}
              interpolation="monotoneX"
            />
          )}
        </VictoryChart>
      )}
    </ChartLayoutLarge>
  );
};

export default ZigChart;
