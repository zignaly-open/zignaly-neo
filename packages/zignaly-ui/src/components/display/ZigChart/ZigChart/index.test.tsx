import React from "react";
import ZigChart from ".";
import { renderWithProvidersUi } from "../../../../utils/testConfig";
import { Box } from "@mui/system";

const mockChartData = [10, 20, 25, 15, 35];

const mockEvents = [
  { x: 1, label: "Event 1" },
  { x: 2, label: "Event 2" },
];

describe("ZigChart component tests", () => {
  it("renders ZigChart component with data", () => {
    const { container } = renderWithProvidersUi(
      <Box width={"500px"}>
        <ZigChart
          id="test-chart"
          data={mockChartData}
          yAxisFormatter={(value) => value.toString()}
          tooltipFormatter={(datum) => `${datum.x}\n${datum.y}`}
          events={mockEvents}
        />
      </Box>,
    );

    expect(container.querySelector("#test-chart")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
