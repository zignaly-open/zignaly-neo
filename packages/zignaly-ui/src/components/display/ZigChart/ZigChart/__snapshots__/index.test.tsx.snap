// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ZigChart component tests renders ZigChart component with data 1`] = `
<div>
  <div
    class="MuiBox-root css-qxh7fn"
  >
    <div
      class="css-vl79hd"
      id="test-chart"
    >
      0
    </div>
  </div>
</div>
`;
